marimo/_runtime/primitives.py

```diff
@@ -0,0 +1,205 @@
+# Copyright 2024 Marimo. All rights reserved.
+from __future__ import annotations
+
+import inspect
+import numbers
+import sys
+import weakref
+from typing import TYPE_CHECKING, Any, Callable, Optional, Union
+
+from marimo._ast.visitor import Name, VariableData
+
+if TYPE_CHECKING:
+    from marimo._runtime.dataflow import DirectedGraph
+
+PRIMITIVES: tuple[type, ...] = (str, numbers.Number, type(None))
+# Weakref instances should be disassociated from related references, as should
+# other "primitives" as they are results and hopefully not hiding some scoped
+# reference.
+CLONE_PRIMITIVES = (weakref.ref,) + PRIMITIVES
+
+if sys.version_info < (3, 9):
+    # Future does not seem to work for this in CI.
+    FN_CACHE_TYPE = Optional["dict[Union[Callable[..., Any], type], bool]"]
+else:
+    FN_CACHE_TYPE = Optional[dict[Union[Callable[..., Any], type], bool]]
+
+
+def is_external(value: Any) -> bool:
+    return "_marimo__cell_" not in inspect.getfile(value)
+
+
+def is_primitive(value: Any) -> bool:
+    # Tuples don't allow for write access
+    if isinstance(value, tuple):
+        return all(map(is_primitive, value))
+    return isinstance(value, PRIMITIVES)
+
+
+def is_primitive_type(value: type) -> bool:
+    return any(issubclass(value, primitive) for primitive in PRIMITIVES)
+
+
+def is_clone_primitive(value: Any) -> bool:
+    return isinstance(value, CLONE_PRIMITIVES)
+
+
+def is_data_primitive(value: Any) -> bool:
+    if is_primitive(value):
+        return True
+
+    # If a numpy array, ensure that it's not an object array.
+    if hasattr(value, "__array_interface__"):
+        dtype = getattr(value, "dtype", None)
+        if dtype is not None and dtype.hasobject:
```

Comment:
Since this is duck-typed, we should have a hasattr:

`if dtype is not None and hasattr(dtype, 'hasobject') and ...`